### ISSUE TITLE
Single-section volumes

### ISSUE THREAD

**Issue by PaulRambags** [mtmse/dotify.formatter.impl#6]
As Dedicon, I would like to have an option to indicate that each volume of the resulting PEF must have exactly one section, so that there are no empty white pages between sections in a duplex pagination.

In the current implementation, a volume can have several sections. The PEF specification is such that in duplex mode a section always starts on a recto page. For instance, if the PEF has this structure:

```
<pef>
  <head>...</head>
  <body>
    <volume duplex="true">
      <section>
        <page>...</page>
      </section>
      <section>
        <page>...</page>
      </section>
    </volume>
  </body>
</pef>
```

then there will be an empty page in between the two `<page>`s. Also if both sections use the same page number counter, the page numbers will be 1 and 3, respectively.

As Dedicon I would like to be able to remove those empty pages between sections, such that the above example becomes:

```
<pef>
  <head>...</head>
  <body>
    <volume duplex="true">
      <section>
        <page>...</page>
        <page>...</page>
      </section>
    </volume>
  </body>
</pef>
```

The page numbers should also be corrected.

Implementation detail:
A volume consists of a pre-content, a body, and a post-content, which are section lists. The pre-content and the post-content can be empty, the body will always have at least one section. Each section has _section properties_ and these may differ per section. The implementation should generate a single section for every volume with the section properties of the first body section.

**Comment by bertfrees** [comment]
Adding an option to do this within Dotify is in theory possible, because OBFL doesn't mention anywhere that a `sequence` or `toc-sequence` or `dynamic-sequence` needs to start on a new sheet (I think, we'd have to check to be sure). However OBFL does have the concept of sheets, and when generating pages, Dotify must know whether it is currently generating a left- or a right-hand page. At first I thought this was no so trivial to implement without starting a new section after the `pre-content`, but maybe it's just a matter of storing the size of the `pre-content` in the `CrossReferenceHandler` (something similar is already done, see the `Overhead` class).

A very simple workaround that does not require any changes in Dotify is to simply not select duplex, and then post-process the PEF before sending it to the printer. The downside if of course that some sheet-related things will not work, like e.g. `break-before="sheet"`, or breaking a volume on a left-hand page.

Related issue: https://github.com/braillespecs/obfl/issues/23

**Comment by PaulRambags** [comment]
>OBFL doesn't mention anywhere that a sequence or toc-sequence or dynamic-sequence needs to start on a new sheet (I think, we'd have to check to be sure

Yes, OBFL mentions that. Quoted from the [ OBFL specification](https://braillespecs.github.io/obfl/obfl-specification.html#L5024):

>Duplex
In a sequence where duplex is enabled, both sides of sheets are counted and paginated. However, if the last sheet of a duplex sequence only has content on the front side, the back side is counted, but not paginated.

In other words, in duplex mode a sequence with an odd number of pages will have an additional empty page, which is counted.

**Comment by bertfrees** [comment]
Hmm, so this means that this new Dotify option would either overwrite what is specified in the OBFL file (which is a bit risky), or we need to change the spec?

The more I think about it the more I think the spec should support this. What do you think about the following proposal?

We already introduced the `break-before` attribute on `sequence` to support forced volume breaks (value "volume"). Why not add the new values "page" and "sheet" (the latter being the default and equivalent to "auto")? In addition we add a `break-before` attribute on `dynamic-sequence` and `toc-sequence`, and also a `break-after` attribute on `sequence`, `dynamic-sequence` and `toc-sequence`.

**Comment by PaulRambags** [comment]
The point is that the PEF format is such that when a volume has more than one section, and a section has an odd number of pages, there will be an empty page inserted. So I need a solution that guarantees that a volume will have no more than one section.

**Comment by bertfrees** [comment]
Yes I understand. But OBFL does not say anything about what the produced PEF should look like (in fact it does not even require that the output is PEF). Dotify can choose whether it creates new `section`s or not. Your only guarantee is that the document will be rendered as specified in the OBFL file, and that is all that matters.

**Comment by bertfrees** [comment]
I'm adding a few examples to clarify things (click to expand).

<details>
<summary>When a <code>toc-sequence</code> is followed by a regular <code>sequence</code>, both are duplex, and we want no new sheet after the TOC:</summary>

```xml
<obfl>
   <layout-master name="default" duplex="true">
      ...
   </layout-master>
   <volume-template>
      <pre-content>
         <toc-sequence master="default">
            ...
         </toc-sequence>
      </pre-content>
   </volume-template>
   <sequence master="default" break-before="page">
      ...
   </sequence>
</obfl>
```

This could result in the following PEF:

```xml
<pef>
   <body>
      <volume duplex="true">
         <section>
            <page>
               <!-- toc -->
            </page>
            <page>
               <!-- body -->
            </page>
         </section>
      </volume>
   </body>
</pef>
```

</details>

<details>
<summary>If you set <code>break-before="sheet"</code> (or you omit <code>break-before</code>), the behavior is as before, i.e. a new PEF <code>section</code> is created for every new sequence:</summary>

```xml
<obfl>
   <layout-master name="default" duplex="true">
      ...
   </layout-master>
   <volume-template>
      <pre-content>
         <toc-sequence master="default">
            ...
         </toc-sequence>
      </pre-content>
   </volume-template>
   <sequence master="default" break-before="sheet">
      ...
   </sequence>
</obfl>
```

```xml
<pef>
   <body>
      <volume duplex="true">
         <section>
            <page>
               <!-- toc -->
            </page>
         </section>
         <section>
            <page>
               <!-- body -->
            </page>
         </section>
      </volume>
   </body>
</pef>
```

But another option could be to insert empty <code>page</code> elements. This PEF and the one above are equivalent:

```xml
<pef>
   <body>
      <volume duplex="true">
         <section>
            <page>
               <!-- toc -->
            </page>
            <page/>
            <page>
               <!-- body -->
            </page>
         </section>
      </volume>
   </body>
</pef>
```

</details>

<details>
<summary>If there is a transition from duplex to simplex or visa-versa, we have to create a new section in the PEF, so the <code>break-before="sheet"</code> has no effect:</summary>

```xml
<obfl>
   <layout-master name="duplex" duplex="true">
      ...
   </layout-master>
   <layout-master name="simplex" duplex="false">
      ...
   </layout-master>
   <volume-template>
      <pre-content>
         <toc-sequence master="duplex">
            ...
         </toc-sequence>
      </pre-content>
   </volume-template>
   <sequence master="simplex" break-before="page">
      ...
   </sequence>
</obfl>
```

```xml
<pef>
   <body>
      <volume duplex="true">
         <section>
            <page>
               <!-- toc -->
            </page>
         </section>
         <section duplex="false">
            <page>
               <!-- body -->
            </page>
         </section>
      </volume>
   </body>
</pef>
```

</details>

**Comment by PaulRambags** [comment]
@bertfrees I like your concept of `break-before="page"`. 👍 
At your last example, you probably mean "`break-before="page"` instead of `break-before="sheet"`.
By the way, this discussion should have been at the OBFL repository, not here.

**Comment by bertfrees** [comment]
> At your last example, you probably mean `"break-before="page"`

Correct
